### PR TITLE
Assignment 5 : OOPS

### PR DESCRIPTION
**Pull request by shafiq005-solin**
# Description

In this assignment, there is parent abstract class which has some properties, methods and some abstract methods. And there are two different classes which are child class of the Parent class. Those classes will override the parent class methods. And create some new functionalities to the methods. 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/assignment-5/BankingSystem/SavingAccount.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        {
+        }
+
+        /// <summary>
+        /// Withdraw the amount from the account with restrictions
+        /// </summary>
+        /// <param name="amount">withdraw amount</param>
+        public override void Withdraw(decimal amount)
+        {
+            if ((this.Balance - amount) < MINIMUMBALANCE)
```
Comment (MithunSubramani) [inline src/assignment-5/BankingSystem/SavingAccount.cs:30]: Will it be ok, if you pass remaining balance as a parameter to the virtual method?

**Review by MithunSubramani — CHANGES_REQUESTED** [review]

File: src/assignment-5/ShapeHierarchy/Circle.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        /// <value> Integer </value>
+        public int Radius { get; set; }
+
+        /// <summary>
+        /// Calculates the area of the Circle
+        /// </summary>
+        /// <returns> Area of the circle </returns>
+        public override int CalculateArea()
+        {
+            return (int)(this.Radius * 2 * 3.14f);
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/Circle.cs:35]: Area was converted into float

File: src/assignment-5/ShapeHierarchy/Circle.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+        /// <summary>
+        /// Display the properties and their values
+        /// </summary>
+        public override void PrintDetails()
+        {
+            Console.WriteLine($"\n\n\t\tCircle\nRadius : {this.Radius}");
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/Circle.cs:43]: Unit was included for both

File: src/assignment-5/EmployeeHierarchy/ValidInput.cs
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/EmployeeHierarchy/ValidInput.cs:1]: Name was converted into ValidateInput.cs

File: src/assignment-5/EmployeeHierarchy/Employee.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        /// Gets or Sets the Name of the employee
+        /// </summary>
+        /// <value> String </value>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or Sets the salary of the employee
+        /// </summary>
+        /// <value> Integer </value>
+        public int Salary { get; set; }
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/EmployeeHierarchy/Employee.cs:33]: Salary converted into Decimal

File: src/assignment-5/ShapeHierarchy/Program.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                Console.Write("Invalid choice..\nTry again : ");
+                choice = ValidInput.GetValidInteger();
+            }
+
+            Console.Clear();
+            switch (choice)
+            {
+                case 1:
+                    {
+                        Console.Write("\t\t Rectangle\nEnter Length : ");
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/Program.cs:44]: Code moved to a private method

File: src/assignment-5/ShapeHierarchy/Program.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                        string color = ValidInput.GetValidString();
+                        Rectangle rectangle = new Rectangle(length, width, color);
+                        Console.WriteLine($"\n\nThe area of the Rectangle : {rectangle.CalculateArea()}");
+                        rectangle.PrintDetails();
+                        break;
+                    }
+
+                case 2:
+                    {
+                        Console.Write("\t\t Circle\nEnter Radius : ");
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/Program.cs:58]: I tried but I couldn't. It is quite complex to do that.

File: src/assignment-5/EmployeeHierarchy/Manager.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        /// <summary>
+        /// Display the properties and their values
+        /// </summary>
+        public override void PrintDetails()
+        {
+            Console.WriteLine("\n\n\t\t Manager details");
+            base.PrintDetails();
+            Console.WriteLine($"Position  : Manager");
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/EmployeeHierarchy/Manager.cs:40]: This was in the given assignment. We should override the method in the child class

File: src/assignment-5/EmployeeHierarchy/Program.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                Console.Write("Invalid choice..\nTry again : ");
+                choice = ValidInput.GetValidInteger();
+            }
+
+            Console.Clear();
+            switch (choice)
+            {
+                case 1:
+                    {
+                        Console.Write("\t\t Manager\nEnter Name : ");
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/EmployeeHierarchy/Program.cs:44]: These code was moved to another private method

File: src/assignment-5/EmployeeHierarchy/Program.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                case 2:
+                    {
+                        Console.Write("\t\t Developer\nEnter Name : ");
+                        string name = ValidInput.GetValidString();
+                        Console.Write("Enter Salary : ");
+                        int salary = ValidInput.GetValidInteger();
+                        Developer developer = new Developer(name, salary);
+                        Console.WriteLine($"\n\nThe Bonus amount is {developer.CalculateBonus()}");
+                        developer.PrintDetails();
+                        break;
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/EmployeeHierarchy/Program.cs:63]: I tried but I couldn't. It is quite complex to do that.

File: src/assignment-5/ShapeHierarchy/ValidInput.cs
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/ValidInput.cs:1]: Name was changed to ValidateInput.cs

[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/ShapeHierarchy/ValidInput.cs:1]: New project was created as Utilities for all validations

File: src/assignment-5/BankingSystem/BankAccount.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            this.Balance += amount;
+        }
+
+        /// <summary>
+        /// Withdraw the amount from the account
+        /// </summary>
+        /// <param name="amount">Withdraw amount</param>
+        public virtual void Withdraw(decimal amount)
+        {
+            if (this.Balance - amount > -1)
```
[the remark this answers is not in the document]
Comment (shafiq005-solin) [reply src/assignment-5/BankingSystem/BankAccount.cs:50]: converted to zero

File: src/assignment-5/BankingSystem/SavingAccount.cs (hunk 1 above)
Comment (shafiq005-solin) [reply src/assignment-5/BankingSystem/SavingAccount.cs:30]: Code was changed now, it is not needed there.

File: src/assignment-5/Utilities/ValidateInput.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            while (string.IsNullOrWhiteSpace(input) || !regex.IsMatch(input))
+            {
+                Console.Write("Invalid String..\nTry again : ");
+                input = Console.ReadLine();
+            }
+
+            return input;
+        }
+
+        /// <summary>
```
Comment (MithunSubramani) [inline src/assignment-5/Utilities/ValidateInput.cs:46]: Can you merge the `GetValidDecimal` and `GetValidInteger` methods, if possible.

File: src/assignment-5/Utilities/Program.cs
Comment (MithunSubramani) [inline src/assignment-5/Utilities/Program.cs:1]: Is `Program.cs` needed?

**Review by MithunSubramani — CHANGES_REQUESTED** [review]

File: src/assignment-5/Utilities/Program.cs
Comment (shafiq005-solin) [reply src/assignment-5/Utilities/Program.cs:1]: Without Program.cs It is not working as excepted

File: src/assignment-5/Utilities/ValidateInput.cs (hunk 11 above)
Comment (shafiq005-solin) [reply src/assignment-5/Utilities/ValidateInput.cs:31]: Both methods were merged and make into a single method

File: src/assignment-5/BankingSystem/Operations.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    public class Operations
+    {
+        /// <summary>
+        /// Handles the SavingAccount and CheckingAccount classes
+        /// </summary>
+        public static void UserInterface()
+        {
+            Console.WriteLine("There are two Accounts available\n1. Saving Account\t\t2.Checking Account");
+            Console.Write("Choose any one (1 or 2): ");
+            int choice = Utilities.ValidateInput.GetValidinput<int>();
```
Comment (MithunSubramani) [inline src/assignment-5/BankingSystem/Operations.cs:19]: Import `Utilities' and use.

**Review by MithunSubramani — APPROVED** [review]

File: src/assignment-5/BankingSystem/Operations.cs (hunk 12 above)
Comment (shafiq005-solin) [reply src/assignment-5/BankingSystem/Operations.cs]: Utilities class imported.

**Pull request merged**